node/src/accountant/scanners/mod.rs

```diff
@@ -938,13 +939,20 @@ impl ReceivableScanner {
                     ),
                 }
                 match txn.commit() {
-                    Ok(_) => {
```

Comment:
Nope. You did not understand my objections. I cannot comment on the exact line but I will leave it here. 

You must not have an operation setting the start_block at any other place than here and only here, that is, after making sure any inbound payments are properly recorded. Now go and look into your code in `node/src/blockchain/blockchain_bridge.rs` at the method `handle_retrieve_transactions`. You'll see what I'm telling is not desirable. You must throw away the `set_start_block` functionality. Which will hopefully also affect some tests.

However that's not all, after it, you need to take care of this area in turn. I noticed that the actor message which introduces the operations in this big function, has the field `new_start_block` declared as an `Option` of `u64`. Please check out the code in blockchain_bridge and look for where you feed this field with values. There is just one place and it always uses Some() wrapping the actual block number around. It's constantly in use. That's apparently wrong. 

We can reason that at the time when we're constructing the `ReceivedPayments` msg there we must already have the start_block defined, which is implied by the preceding operations in there (that you incorporate by one of your past cards). All I'm saying is that field does not need to be optional. However, I'm tired to death at the moment and I'd want you to double-check if I did the survey right. 

Once we have an agreement about it, we need to move over here and fix it so that `if let Some(new_start_block) = msg.new_start_block {` makes sense in this new perspective. The field is not optional anymore. So what can we do? Basically, we've got rid off one of the else clauses and you can erase the lines 953-955. That's great news.

Reply:
I agree setting the start block number in a single place is the right way to go.

Where is the single best place for calling `set_start_block`? Let me ask a few questions to show my thought process to arrive at blockchain_bridge.

Which _impl_ 'knows' the most about the block number and the range of blocks we're allowed to query with each call? Doesn't the receivable scanner only discover the block numbers where payments exist and those numbers might not be in the final block scanned? If we remove setting the start block from `blockchain_bridge`, what happens if we receive no payments? How will blockchain bridge avoid scanning the same blocks again and again? As the latest block number increases while our start_block_number remains unchanged, isn't there risk that over-time the block range would increase beyond what is allowed by the blockchain service before we receive a payment?

The fundamental requirement of GH-606 is to allow the `start_block_number` to be dynamic and not static in a new database. Perhaps I chose poorly with the implementation, but to me a dynamic value starts as an undefined number -- `null` in the database and an `Option` in the code. `None` is more difficult to misinterpret than `0u64` (u64 can't be negative otherwise `-1` might be just as difficult to misinterpret). `start_block_nbr` immediately starts with `None` and updates it to `Some` with the latest block number from the blockchain. After the database is initialized `start_block_nbr` is always `Some`. Could it be u64::MIN? Yes. Do you prefer this?

Reply:
Ok. If I could strike through the previous comments I would, but GitHub markdown doesn't support it. This has percolated in my subconscious and I realize I am wrong and you are right. All you had to mention is that we have to commit the `new_start_block` in the same transaction with the `ReceivedPayments`. I will also make the `ReceivedPayments` message have `new_start_block: u64` no `Option` needed.